### PR TITLE
Avoid a few array allocations with DefaultBinder

### PR DESCRIPTION
**Pull request by stephentoub**
In places where we're not handing out a `T[]` and thus actually need a `T[]`, we can instead just get a span from the `ListBuilder<T>` being used.

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs [hunk 1]
```diff
@@ -967,5 +967,102 @@ private CheckValueStatus TryChangeType(ref object? value, ref bool copyBack)
 
             return CheckValueStatus.ArgumentException;
         }
+
+        // Helper to build lists of MemberInfos. Special cased to avoid allocations for lists of one element.
```
Comment (jkotas) [inline src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs:971]: Move this into a separate .cs file and de-dup it with the nativeaot copy?

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs:971]: Sure. Do we want to keep the nativeaot's DEBUG-only logic?

Comment (jkotas) [reply src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs:971]: Either way is fine with me.

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/RuntimeType.cs:971]: Subsequently we might also want to dedup ListBuilder and ValueListBuilder. The main thing the former provides is a non-allocating slot for a single entry, which ValueListBuilder can also achieve via the scratch span passed in, and that can be created now via an inlinearray, a collection expression that itself uses an inlinearray, or a span wrapping a single local.

**Review by jkotas — APPROVED** [review]
Thanks

**Pull request merged**